bioframe/genomeops.py

```diff
@@ -295,3 +295,76 @@ def frac_gene_coverage(df, ucsc_mrna):
     df_gene_coverage = ops.count_overlaps(df_gene_coverage, mrna)
 
     return df_gene_coverage
+
+
+def pair_by_distance(df, min_sep, max_sep, from_ends=False, suffixes=("_1", "_2")):
+    """
+    From a dataframe of genomic intervals, find all unique pairs of intervals
+    that are between ``min_sep`` and ``max_sep`` bp separated from each other.
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        A BED-like dataframe.
+    min_sep, max_sep : int
+        Minimum and maximum separation between intervals in bp.
+    from_ends : bool, optional, default: False
+        Calculate distances between interval endpoints. If False, distances are
+        calculated between interval midpoints (default).
+    suffixes : (str, str), optional
+        The column name suffixes for the two interval sets in the output.
+        The first interval of each output pair is always upstream of the
+        second.
+
+    Returns
+    -------
+    pandas.DataFrame
+        A BEDPE-like dataframe of paired intervals from ``df``.
+
+    """
+    if min_sep >= max_sep:
+        raise ValueError("min_sep must be less than max_sep")
+
+    mids = (df["start"] + df["end"]) // 2
+
+    # For each interval, generate a probe interval on its right
+    if from_ends:
+        ref = df["end"]
+    else:
+        ref = mids
+    right_probe = df[["chrom"]].copy()
+    right_probe["start"] = ref + min_sep // 2
+    right_probe["end"] = ref + int(np.ceil(max_sep / 2))
+
+    # For each interval, also generate a probe interval on its left
+    if from_ends:
+        ref = df["start"]
+    else:
+        ref = mids
+    left_probe = df[["chrom"]].copy()
+    left_probe["start"] = ref - max_sep // 2
+    left_probe["end"] = ref - int(np.ceil(min_sep / 2))
+
+    # Ignore probes that fall out of bounds
+    mask = (left_probe["start"] > 0) & (right_probe["start"] > 0)
```

Comment:
hmm, I'm trying to understand this.
Say, we have two locations, 
refs=[1,3]
min_sep=1,
max_sep=10

right_probe["start"] = ref + min_sep // 2 = [1,3]
right_probe["end"] = ref + int(np.ceil(max_sep / 2)) = [1,3]+5=[6,8]
left_probe["start"] = ref - max_sep // 2 = [1,3] - 5 = [-4,-2]
left_probe["end"] = ref - int(np.ceil(min_sep / 2)) = [1,3]-1=[0,2]

the mask will be [False,False], which seems wrong?..

Reply:
Good point. I think that filter was in the bedtools implementation because it would choke on negative coordinate inputs. There shouldn't be a problem with `ops.overlap`, right?

Reply:
Yeah, seems to work:
![image](https://user-images.githubusercontent.com/912906/101140554-6adbcb80-3613-11eb-9f81-6e3ededb1bf2.png)